// appium-doctor/build/lib/demo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FileCheck = exports.DirCheck = void 0;

require("source-map-support/register");

var _utils = require("./utils");

var _appiumSupport = require("appium-support");

var _teen_process = require("teen_process");

var _doctor = require("./doctor");

var _logger = _interopRequireDefault(require("./logger"));

var _prompt = require("./prompt");

let checks = [];

class DirCheck extends _doctor.DoctorCheck {
  constructor(path) {
    super({
      autofix: false
    });
    this.path = path;
  }

  async diagnose() {
    if (!(await _appiumSupport.fs.exists(this.path))) {
      return (0, _utils.nok)(`Could NOT find directory at '${this.path}'!`);
    }

    let stats = await _appiumSupport.fs.lstat(this.path);
    return stats.isDirectory() ? (0, _utils.ok)(`Found directory at: ${this.path}`) : (0, _utils.nok)(`'${this.path}' is NOT a directory!`);
  }

  async fix() {
    return `Manually create a directory at: ${this.path}`;
  }

}

exports.DirCheck = DirCheck;
checks.push(new DirCheck('/tmp/appium-doctor'));
checks.push(new DirCheck('/tmp/appium-doctor/demo'));

class FileCheck extends _doctor.DoctorCheck {
  constructor(path) {
    super({
      autofix: true
    });
    this.path = path;
  }

  async diagnose() {
    return (await _appiumSupport.fs.exists(this.path)) ? (0, _utils.ok)(`Found file at: ${this.path}`) : (0, _utils.nok)(`Could NOT find file at '${this.path}'!`);
  }

  async fix() {
    _logger.default.info(`The following command need be executed: touch '${this.path}'`);

    let yesno = await (0, _prompt.fixIt)();

    if (yesno === 'yes') {
      await (0, _teen_process.exec)('touch', [this.path]);
    } else {
      _logger.default.info(`Skipping you will need to touch '${this.path}' manually.`);

      throw new _doctor.FixSkippedError('bbb');
    }
  }

}

exports.FileCheck = FileCheck;
checks.push(new FileCheck('/tmp/appium-doctor/demo/apple.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/pear.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/orange.fruit'));
var _default = checks;
exports.default = _default;require('source-map-support').install();
