// appium-doctor/build/lib/dev.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.BinaryIsInPathCheck = exports.AndroidSdkExists = void 0;

require("source-map-support/register");

var _doctor = require("./doctor");

var _utils = require("./utils");

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

require("@colors/colors");

let checks = [];

class BinaryIsInPathCheck extends _doctor.DoctorCheck {
  constructor(binary) {
    super();
    this.binary = binary;
  }

  async diagnose() {
    const resolvedPath = await (0, _utils.resolveExecutablePath)(this.binary);

    if (!resolvedPath) {
      return (0, _utils.nok)(`${this.binary} is MISSING in PATH: ${process.env.PATH}`);
    }

    return (0, _utils.ok)(`${this.binary} was found at ${resolvedPath}`);
  }

  fix() {
    return `Manually install the ${this.binary.bold} binary and add it to ${'PATH'.bold}.`;
  }

}

exports.BinaryIsInPathCheck = BinaryIsInPathCheck;
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'mvn.bat' : 'mvn'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'ant.bat' : 'ant'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'adb.exe' : 'adb'));

class AndroidSdkExists extends _doctor.DoctorCheck {
  constructor(sdk) {
    super();
    this.sdk = sdk;
  }

  async diagnose() {
    if (typeof process.env.ANDROID_HOME === 'undefined') {
      return (0, _utils.nok)(`${this.sdk} could not be found because ANDROID_HOME is NOT set!`);
    }

    let sdkPath = _path.default.resolve(process.env.ANDROID_HOME, _path.default.join('platforms', this.sdk));

    return (await _appiumSupport.fs.exists(sdkPath)) ? (0, _utils.ok)(`${this.sdk} was found at: ${sdkPath}`) : (0, _utils.nok)(`${this.sdk} could NOT be found at '${sdkPath}'!`);
  }

  fix() {
    if (typeof process.env.ANDROID_HOME === 'undefined') {
      return `Manually configure ${'ANDROID_HOME'.bold}.`;
    }

    return `Manually install the ${this.sdk.bold} sdk.`;
  }

}

exports.AndroidSdkExists = AndroidSdkExists;
checks.push(new AndroidSdkExists('android-16'));
checks.push(new AndroidSdkExists('android-19'));
var _default = checks;
exports.default = _default;require('source-map-support').install();
