// appium-doctor/build/lib/prompt.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clear = clear;
exports.configure = configure;
exports.fixIt = fixIt;

require("source-map-support/register");

var _utils = require("./utils");

let persistentResponse;
const fixItQuestion = {
  type: 'list',
  name: 'confirmation',
  message: 'Fix it:',
  choices: ['yes', 'no', 'always', 'never'],

  filter(val) {
    return val.toLowerCase();
  }

};

function configure(opts) {
  if (opts.yes) {
    persistentResponse = 'yes';
  }

  if (opts.no) {
    persistentResponse = 'no';
  }
}

function clear() {
  persistentResponse = undefined;
}

async function fixIt() {
  if (persistentResponse) {
    return persistentResponse;
  }

  let resp = await _utils.inquirer.prompt(fixItQuestion);
  persistentResponse = resp.confirmation === 'always' ? 'yes' : persistentResponse;
  persistentResponse = resp.confirmation === 'never' ? 'no' : persistentResponse;
  return persistentResponse || resp.confirmation;
}require('source-map-support').install();
